lib/src/v3/connection.dart

```diff
@@ -73,40 +73,61 @@ abstract class _PgSessionBase implements PgSession {
 
   PgConnectionImplementation get _connection;
 
+  /// Runs [callback], guarded by [_operationLock] and cleans up the pending
+  /// resource afterwards.
+  Future<T> _withResource<T>(FutureOr<T> Function() callback) {
+    return _operationLock.withResource(() async {
+      assert(_connection._pending == null);
+
+      try {
+        return await callback();
+      } finally {
+        _connection._pending = null;
+      }
+    });
+  }
+
   /// Sends a message to the server and waits for a response [T], gracefully
   /// handling error messages that might come in instead.
   Future<T> _sendAndWaitForQuery<T extends ServerMessage>(ClientMessage send) {
     final trace = StackTrace.current;
 
-    return _operationLock.withResource(() {
+    return _withResource(() {
       _connection._channel.sink
           .add(AggregatedClientMessage([send, const SyncMessage()]));
 
-      final completer = Completer<T>();
-      final syncComplete = Completer<void>.sync();
+      final doneWithOperation = Completer<void>.sync();
+      Result<T>? result;
 
       _connection._pending = _CallbackOperation(_connection, (message) async {
         if (message is T) {
-          completer.complete(message);
+          result = Result.value(message);
         } else if (message is ErrorResponseMessage) {
-          completer.completeError(
-              PostgreSQLException.fromFields(message.fields), trace);
-        } else if (message is ReadyForQueryMessage) {
-          if (!completer.isCompleted) {
-            completer.completeError(
-                StateError('Operation did not complete'), trace);
-          }
+          final exception = PostgreSQLException.fromFields(message.fields);
 
-          syncComplete.complete();
+          result = Result.error(exception, trace);
+
+          // If the error is severe enough for the operation or the whole
+          // connection to abort, we should also release the lock.
+          if (exception.willAbortCommand) {
+            doneWithOperation.complete();
+          }
+        } else if (message is ReadyForQueryMessage) {
+          // This is the message we've been waiting for, the server is signalling
+          // that it's ready for another message - so we can release the lock.
+          doneWithOperation.complete();
         } else {
-          completer.completeError(
-              StateError('Unexpected message $message'), trace);
+          result =
+              Result.error(StateError('Unexpected message $message'), trace);
```

Comment:
`doneWithOperation` seems to be not completed in this case. Wouldn't it hang in this case?

Reply:
The tests are now stuck for 4h+ hours running... please check if this is related to that.

Reply:
Sorry about the timeouts, I couldn't reproduce them with podman somehow but they happen with Docker... They should be fixed now, but I've also reduced the timeout for the whole CI to catch them earlier.

If we get an unexpected message, there's either a serious bug in the v3 implementation or something very wrong the postgres connection. I think the best course might be to just close the connection in that case, we can't really expect to recover if we have no idea what's going on.